README.md

```diff
@@ -0,0 +1,182 @@
+| auth      | string | `null` - (no auth) | Optional |
+| domain    | string | firebaseio.com     | Optional |
+| debug     | bool   | `true`             | Optional - set to `false` to surpress error logging within the Firebase class |
+
+The domain and instance are used to construct the url requests are made against in the following was: https://{instance}.{domain}
+
+```squirrel
+#require "Firebase.class.nut:1.0.0"
+
+const FIREBASE_AUTH_KEY = "<-- Your Firebase Auth Key -->"
+
+firebase <- Firebase("myfirebase", FIREBASE_AUTH_KEY)
+```
+
+## Class Methods
+
+### on(*path, callback*)
+
+Listens for changes at a particular location (*path*) or a node below *path*. When changes are detected, the callback method will be invoked. The callback method takes two parameters: *path* and *data*.
+
+The *path* parameter returns the full path of the node that was modified (this allows you to determin if the root of the path you're tracking changed, or if a node below it changed).
+
+The *data* parameter contains the modified data.
+
+```squirrel
+// Register a callback handler
+firebase.on("/settings", function(path, data) {
+    server.log("Changes detected at: " + path);
+    server.log("New Data: " + http.jsonencode(data));
+
+    // Make sure the data we're looking for exists
+    if ("color" in data) {
+        // Send the new settings to the device
+        device.send("newColor", data.color);
+    }
+});
+```
+
+### stream(*[path, onErrorCallback]*)
+
+Creates a streaming request using the *path* as the base address to track. If no *path* is supplied, the root of the instance ("/") will be used.
+
+An optional onErrorCallback parameter can be supplied that will be invoked if errors occur while making streaming requests.
+
+If no onErrorCallback is not supplied, the Firebase class will attempt to silently and automatically reconnect when it encounters an error. If an onErrorCallback is supplied, it is up to the developer to re-initiate the steam() request in the onErrorCallback.
+
+The onErrorCallback takes a single parameter - the [HTTP Response Table](https://electricimp.com/docs/api/httprequest/sendasync/) from the request.
+
+```squirrel
+// Setup onError handler
+function onStreamError(resp) {
+    server.error("Firebase encountered and error:");
+    server.error(resp.statuscode + " - " + resp.body);
+    imp.wakeup(1, openStream });
+}
+
+// Wrap up the process of opening a stream
+function openStream() {
+    firebase.stream("/", onStreamError);
+}
+
+openStream();
+```
+
+### isStreaming()
+Returns `true` is the stream is currently open, and `false` otherwise.
+
+### closeStream()
+Closes the current stream (if it is open), and takes no action otherwise.
+
+### fromCache(*[path]*)
+Reads the local/cached copy of the data at the specified path - intended for use in a .on() handler to prevent unnecessary communication with Firebase through the *read* method.
+
+```squirrel
+firebase.stream("/settings", function(path, data) {
```

Comment:
this also works, and shows that the cache lookup is not relative to the stream.

    firebase.stream("/");
    firebase.on("/settings", function(path, data) {
        // Check if the location setting changed
        if ("location" in data || path.find("location")) {
            // if it did, grab the location data from Firebase
            // use fromCache instead of read to avoid an unnecessary web request
            local location = firebase.fromCache("/settings/location");
            
            // Send the new location information to the device
            device.send("updateLocation", location);
        }
    });